### PR TITLE
Adding GPU automatic mixed precision training

### PR DESCRIPTION
**Pull request by vinhngx**
Automatic Mixed Precision (AMP) training on GPU for TensorFlow has been recently introduced:

https://medium.com/tensorflow/automatic-mixed-precision-in-tensorflow-for-faster-ai-training-on-nvidia-gpus-6033234b2540

Automatic mixed precision training makes use of both FP32 and FP16 precisions where appropriate. FP16 operations can leverage the Tensor cores on NVIDIA GPUs (Volta, Turing or newer architectures) for improved throughput. Mixed precision training also often allows larger batch sizes.

This PR adds GPU automatic mixed precision training to `DeepSpeech` via passing the flags value `--auto_mixed_precision=True`.

```
./DeepSpeech.py --train_files ./train.csv --dev_files ./dev.csv --test_files ./test.csv --automatic_mixed_precision
```

On a Volta generation V100 GPU, automatic mixed precision speeds up DeepSpeech training and  evaluation by ~30%-40%.

Without AMP:
```
Epoch 0 |   Training | Elapsed Time: 4:06:02 | Steps: 60592 | Loss: 171.049121
Epoch 0 | Validation | Elapsed Time: 0:11:49 | Steps: 12229 | Loss: 152.989010 | Dataset: ./dev.csv
...
Epoch 4 |   Training | Elapsed Time: 4:06:11 | Steps: 60592 | Loss: 170.192227
Epoch 4 | Validation | Elapsed Time: 0:11:55 | Steps: 12229 | Loss: 151.621022 | Dataset: ./dev.csv
I Early stop triggered as (for last 4 steps) validation loss: 151.621022 with standard deviation: 0.084157 and mean: 151.590638
I FINISHED optimization in 21:30:26.464164
```

With AMP:
```
Epoch 0 |   Training | Elapsed Time: 2:51:47 | Steps: 60592 | Loss: 171.021818
Epoch 0 | Validation | Elapsed Time: 0:06:54 | Steps: 12229 | Loss: 151.720192 | Dataset: ./dev.csv
...
Epoch 3 |   Training | Elapsed Time: 2:51:46 | Steps: 60592 | Loss: 170.514453
Epoch 3 | Validation | Elapsed Time: 0:06:47 | Steps: 12229 | Loss: 151.553078 | Dataset: ./dev.csv
I Saved new best validating model with loss 151.553078 to: ./amp_training/best_dev-242254
I Early stop triggered as (for last 4 steps) validation loss: 151.553078 with standard deviation: 0.029790 and mean: 151.706624
I FINISHED optimization in 11:59:06.455024
```

To learn more about mixed precision and how it works:

   [Overview of Automatic Mixed Precision for Deep Learning](https://developer.nvidia.com/automatic-mixed-precision)
    [NVIDIA Mixed Precision Training Documentation](https://docs.nvidia.com/deeplearning/sdk/mixed-precision-training/index.html)
    [NVIDIA Deep Learning Performance Guide](https://docs.nvidia.com/deeplearning/sdk/dl-performance-guide/index.html)

### CODE REVIEW

File: DeepSpeech.py [hunk 1]
```diff
@@ -446,6 +446,10 @@ def train():
 
     # Building the graph
     optimizer = create_optimizer()
+    
+    # Enable mixed precision training
+    if FLAGS.automatic_mixed_precision:
+        optimizer = tf.train.experimental.enable_mixed_precision_graph_rewrite(optimizer)
```
Comment (lissyx) [inline DeepSpeech.py]: @vinhngx  This is under experimental, could you please elaborate / link to further docs on what one can expect / not expect from this ?

Comment (vinhngx) [reply DeepSpeech.py]: Hi @lissyx, though under experimental, automatic mixed precision has been tested quite extensively and incorporated into the TF official model zoo (https://github.com/tensorflow/models/pull/7419, https://github.com/tensorflow/models/pull/7531), tensor2tensor (https://github.com/tensorflow/tensor2tensor/pull/1680), NVIDIA TF model zoo (https://github.com/NVIDIA/DeepLearningExamples/tree/master/TensorFlow). Most of the time, AMP provides speed up at little or no loss to model accuracy. 

The docstring of this function is a good place of reference:
https://github.com/tensorflow/tensorflow/pull/29249/files

**Comment by reuben** [comment]
Thanks for the PR! Can you restore/continue training a mixed precision training checkpoint with mixed precision training disabled?

And some minor questions:
1. In TF2.0 this API is incompatible with `tf.GradientTape`. Do you know if `tf.GradientTape` will be supported in the near term? We would need it for multi-GPU learning with a custom training loop.
2. Is there any way to know if the API actually did anything? E.g. for logging a diagnostic message.

**Comment by vinhngx** [comment]
Hi @reuben 

> Thanks for the PR! Can you restore/continue training a mixed precision training checkpoint with mixed precision training disabled?

FP32 and AMP checkpoints are incompatible as there are graph edits carried out by graph rewrite, e.g. type casting nodes. So in my experience, it's not possible to cross-load and retrain in the most straightforward way.  

> 
> And some minor questions:
> 
>     1. In TF2.0 this API is incompatible with `tf.GradientTape`. Do you know if `tf.GradientTape` will be supported in the near term? We would need it for multi-GPU learning with a custom training loop.

Yes tf.GradientTape is supported. See an example of CTL in the TF model zoo for NCF https://github.com/tensorflow/models/pull/7436/files

>     2. Is there any way to know if the API actually did anything? E.g. for logging a diagnostic message.

Usually the API prints out some statistics such as how many nodes it has converted. In case of DeepSpeech, I haven't seen such report though. (But when trying to load from an FP32 checkpoint it will complain).

**Comment by reuben** [comment]
In my tests this is changing the training behavior even on my laptop without any GPUs and tensorflow (instead of tensorflow-gpu) installed. Here's the loss development log with and without the flag specified with `./bin/run-ldc39s1.sh`. https://gist.github.com/reuben/fe1d4858f2a98238d244e9db10ad428e

**Comment by vinhngx** [comment]
hi @reuben what data are you trying with. I was trying with the full training set and got largely similar results (i.e. loss) with and w/o AMP on Tensorflow-GPU on a V100:

w/o AMP: `I Early stop triggered as (for last 4 steps) validation loss: 151.621022 with standard deviation: 0.084157 and mean: 151.590638
I FINISHED optimization in 21:30:26.464164`

with AMP: `I Early stop triggered as (for last 4 steps) validation loss: 151.553078 with standard deviation: 0.029790 and mean: 151.706624
I FINISHED optimization in 11:59:06.455024`
